sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,735 @@
+      return String.format(
+          "sf_copy_csv_%s_%s",
+          new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date()),
+          UUID.randomUUID().toString().subSequence(0, 8) // first 8 chars of UUID should be enough
+          );
+    }
+
+    private static class CopyIntoStageFn extends DoFn<Object, String> {
+      private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
+      private final String query;
+      private final String table;
+      private final String integrationName;
+      private final String stagingBucketDir;
+      private final SnowflakeService snowflakeService;
+
+      private CopyIntoStageFn(
+          SerializableFunction<Void, DataSource> dataSourceProviderFn,
+          String query,
+          String table,
+          String integrationName,
+          String stagingBucketDir,
+          SnowflakeService snowflakeService) {
+        this.dataSourceProviderFn = dataSourceProviderFn;
+        this.query = query;
+        this.table = table;
+        this.integrationName = integrationName;
+        this.stagingBucketDir = stagingBucketDir;
+        this.snowflakeService = snowflakeService;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String output =
+            snowflakeService.copyIntoStage(
+                dataSourceProviderFn, query, table, integrationName, stagingBucketDir);
+
+        context.output(output);
+      }
+    }
+
+    public static class MapCsvToStringArrayFn extends DoFn<String, String[]> {
+      @ProcessElement
+      public void processElement(ProcessContext c) throws IOException {
+        String csvLine = c.element();
+        CSVParser parser = new CSVParserBuilder().withQuoteChar(CSV_QUOTE_CHAR.charAt(0)).build();
+        String[] parts = parser.parseLine(csvLine);
+        c.output(parts);
+      }
+    }
+
+    private static class MapStringArrayToUserDataFn<T> extends DoFn<String[], T> {
+      private final CsvMapper<T> csvMapper;
+
+      public MapStringArrayToUserDataFn(CsvMapper<T> csvMapper) {
+        this.csvMapper = csvMapper;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        context.output(csvMapper.mapRow(context.element()));
+      }
+    }
+
+    public static class CleanTmpFilesFromGcsFn extends DoFn<Object, Object> {
+      private final String stagingBucketDir;
+
+      public CleanTmpFilesFromGcsFn(String stagingBucketDir) {
+        this.stagingBucketDir = stagingBucketDir;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws IOException {
+        String combinedPath = stagingBucketDir + "*";
+        List<ResourceId> paths =
+            FileSystems.match(combinedPath).metadata().stream()
+                .map(metadata -> metadata.resourceId())
+                .collect(Collectors.toList());
+
+        FileSystems.delete(paths);
```

Comment:
Probably specify MoveOptions.IGNORE_MISSING_FILES to not fail incase this bundle is retries after a failure (either the source or the destination should exist).
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MoveOptions.java#L30